templates/base/client/.rtag/State.tsx.hbs

```diff
@@ -0,0 +1,160 @@
+import React, { useRef, useLayoutEffect, useContext } from "react";
+import { RtagContext } from "./context";
+import { RtagConnection } from "./client";
+import {
+  {{userState}} as UserState,
+  {{#each types}}
+  {{@key}},
+  {{/each}}
+} from "./types";
+{{#each plugins}}
+import {{this}}Plugin from "../plugins/{{this}}";
+{{/each}}
+
+{{#each plugins}}
+window.customElements.define("{{makePluginName this}}", {{this}}Plugin);
+{{/each}}
+
+type BaseProps<T> = { value: T };
+
+function KVDisplay(props: { label: string; children: JSX.Element }) {
+  return (
+    <div className="kv-display">
+      <span>{props.label}: </span>
+      {props.children},
+    </div>
+  );
+}
+
+function ArrayDisplay<T>(props: { value: T[]; children: (value: T) => JSX.Element }) {
+  return (
+    <span className="array-display">
+      [
+      {props.value.map((val, i) => (
+        <div className="array-item" key={i}>
+          {props.children(val)}
+        </div>
+      ))}
+      ]
+    </span>
+  );
+}
+
+function OptionalDisplay<T>(props: { value: T | undefined; children: (value: T) => JSX.Element }) {
+  return <span className="optional-display">{props.value !== undefined ? props.children(props.value) : ""}</span>;
+}
+
+function EnumDisplay(props: { value: number; enum: object }) {
+  const labels = Object.entries(props.enum)
+    .filter(([_, value]) => typeof value === "number")
+    .map(([label, _]) => label);
+  return <span className="enum-display">{labels[props.value]}</span>;
+}
+
+function StringDisplay({ value }: BaseProps<string>) {
+  return <span className="string-display">"{value}"</span>;
+}
+
+function NumberDisplay({ value }: BaseProps<number>) {
+  return <span className="number-display">{value}</span>;
+}
+
+function BooleanDisplay({ value }: BaseProps<boolean>) {
+  return <span className="boolean-display">{value ? "true" : "false"}</span>;
+}
+
+function PluginDisplay<T>({ value, component }: { value: T; component: string }) {
+  const { state, connection } = useContext(RtagContext)!;
+  const ref = useRef<{ val: T; state: UserState; client: RtagConnection }>();
```

Comment:
it looks like here you initialize ref to undefined and then use the ! to tell the compiler that it is never undefined. am i missing something? if not, you should definitely handle undefined or pass in a defined value to begin with

Reply:
well i guess the ref is initialized by react somehow when I do the `React.createElement(component, { ref })`? i'm not sure how this works